File: Linked list/148. Sort List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution
{
public:
    ListNode *merge(ListNode *left, ListNode *right)
    {
        ListNode *newList = new ListNode(-1);
        ListNode *newHead = newList;

        while (left && right)
        {
            if (left->val < right->val)
            {
                newHead->next = left;
                left = left->next;
            }
            else
            {
                newHead->next = right;
                right = right->next;
            }
            newHead = newHead->next;
        }

        while (left != NULL)
        {
            newHead->next = left;
            left = left->next;
            newHead = newHead->next;
        }
        while (right != NULL)
        {
            newHead->next = right;
            right = right->next;
            newHead = newHead->next;
        }
        return newList->next;
    }

    ListNode *sortList(ListNode *head)
    {
        if (head == NULL || head->next == NULL)
            return head;

        ListNode *temp = head;
        ListNode *slow = head;
        ListNode *fast = head;
        while (fast && fast->next)
        {
            temp = slow;
            slow = slow->next;
            fast = fast->next->next;
        }

        temp->next = NULL;
        ListNode *left = sortList(head);
        ListNode *right = sortList(slow);
        ListNode *newHead = merge(left, right);

        return newHead;
    }
};

File: Linked list/148. Sort List.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode sortList(ListNode head) {
        if(head==null||head.next==null)
            return head;
        ListNode middle=getMiddle(head);
        ListNode midnext=middle.next;
        middle.next=null;
        ListNode left=sortList(head);
        ListNode right=sortList(midnext);
        ListNode sortedList=merge(left,right);
        return sortedList;
    }
    public static ListNode merge(ListNode left,ListNode right)
    {
        ListNode ans=null;
        if(left==null)
            return right;
        if(right==null)
            return left;
        if(left.val<=right.val)
        {
            ans=left;
            ans.next=merge(left.next,right);
        }
        else
        {
            ans=right;
            ans.next=merge(left,right.next);
        }
        return ans;
    }
    public static ListNode getMiddle(ListNode head)
    {
        if(head==null)
            return null;
        ListNode slow=head;
        ListNode fast=head;
        while(fast.next!=null && fast.next.next!=null)
        {
            slow=slow.next;
            fast=fast.next.next;
        }
        return slow;
    }
    
}